File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    answer = 0
    
    
    
    point = 0
    ist = []
    dp = []
    ist.append(N)
    dp.insert(0,ist)
    ist = []


    if N == number :
        answer = 1
    else:
        for i in range(1,8):  # 몇개로 이루어져있는지 개수 i
            for j in range(0,i):   # i개로 이루어져있는 경우 연산하는 경우의 수

                for k in dp[j]:

                    for z in dp[-j-1]:
                        ist.append(k + z)
                        ist.append(k - z)
                        ist.append(k * z)
                        if z != 0:
                            ist.append(k // z)
            ist.append(int(str(N)*(i+1)))
            s = set(ist)
            if number in s:
                point = i
                break
            else:
                dp.insert(i, s)
                ist = []
            # print("dp : ", dp)
        if point != 0:
            answer = point+1
        else:
            answer = -1
    
    
    
    return answer

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        int answer = -1;

        ArrayList<ArrayList<Integer>> dp = new ArrayList<>();
        ArrayList<Integer> temp = new ArrayList<>();
        
        temp.add(N);
        dp.add(temp);
        
        temp = new ArrayList<>();
        temp.add(N+N);
        temp.add(N-N);
        temp.add(N*N);
        temp.add(N/N);
        temp.add(N*11);
        dp.add(temp);
        
        if (N == number) {
            answer = 1;
        } else if (dp.get(dp.size()-1).indexOf(number) != -1) {
            answer = 2;
        } else {
            for(int i = 2; i < 9; i++){
                temp = new ArrayList<>();
                String str = "";
                
                for(int j = 0; j < i; j++){
                    for(int k : dp.get(j)){
                        for(int z : dp.get(i-j-1)){
                            temp.add(z+k);
                            temp.add(z-k);
                            temp.add(z*k);
                            if (z != 0 && k != 0) {
                                temp.add(z/k);
                            }
                        }
                    }
                }
                Set<Integer> set = new HashSet<>(temp);
                temp = new ArrayList<>(set);
                
                for(int j = 0; j <= i; j++){
                    str += Integer.toString(N);
                }
                temp.add(Integer.parseInt(str));
                dp.add(temp);
                
                if (dp.get(dp.size()-1).indexOf(number) != -1) {
                    System.out.println(dp.get(dp.size()-1).indexOf(number));
                    answer = i + 1;
                    return answer;
                }
                    
            }
        }
        return answer;
    }
}